Add/source param to get subscriptions call

KokkieH opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

We want to add a source parameter to WooCommerce.com API requests to the subscriptions endpoint

Related issue on WooCommerce.com: 13928-gh-Automattic/woocommerce.com

### How to test the changes in this Pull Request:

Test My subscriptions

1. Open WooCommerce > Extensions
2. Open the _My Subscriptions_ tab
3. Connect to WooCommerce.com
4. Update your list of subscriptions
5. Confirm that your subscriptions are loaded
6. Confirm that other Helper functions that use the WC_Helper_API::get() and url() methods still works normally, e.g. visit the Plugins page and ensure you can activate, deactivate, and view the View Details modal for plugins hosted on WooCommerce.com

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

KokkieH committed:
Add source param for subscription endpoint calls

- Adds an optional $source argument to the WC_Helper_API::url() method
- Get URL of page making WC_Helper::get-subscripitonts() request and add a
  source arg to the WC_Helper_API::get() request made by that method
- Pass that source arg to the WC_Helper_API::url() methd from
  WC_Helper_API::get()

KokkieH committed:
Add changelog file

andfinally commented on plugins/woocommerce/includes/admin/helper/class-wc-helper-api.php:
```diff
@@ -163,9 +166,9 @@ public static function put( $endpoint, $args = array() ) {
+	public static function url( $endpoint, $source = '' ) {
 		$endpoint = ltrim( $endpoint, '/' );
-		$endpoint = sprintf( '%s/%s', self::$api_base, $endpoint );
+		$endpoint = sprintf( '%s/%s/%s', self::$api_base, $endpoint, $source );
```
I wonder if it's worth trimming off any trailing `/`, in case `$source` is `''`?

```suggestion
		$endpoint = sprintf( '%s/%s/%s', self::$api_base, $endpoint, $source );
		$endpoint = rtrim( $endpoint, '/' );
```

KokkieH replied:
From my testing no request that hits the subscriptions endpoint on WooCommerce.com ends in a `/`, regardless whether or not it includes a query string, but no harm it being cautious, I think.

andfinally commented on plugins/woocommerce/includes/admin/helper/class-wc-helper.php:
```diff
@@ -1270,11 +1270,24 @@ public static function get_subscriptions() {
 			return $data;
 		}
 
+		$source = $_SERVER['REQUEST_URI'];
```
I would name this `$url` or suchlike, and initialise an empty string `$source`, to make it clear that `$source` is something else.

KokkieH replied:
That makes sense. I went with `$request_uri`, as that's ultimately what it is.

andfinally commented on plugins/woocommerce/includes/admin/helper/class-wc-helper.php:
```diff
@@ -1270,11 +1270,24 @@ public static function get_subscriptions() {
 		}
 
+		$source = $_SERVER['REQUEST_URI'];
+		if ( preg_match( '/\S+wc-addons?\S+/', $source ) ):
+			$source = 'my-subscriptions';
+		elseif ( preg_match( '/\S+plugins.php?\S+/', $source ) ):
+			$source = 'plugins';
+		elseif ( preg_match( '/\S+wc-admin?\S+/', $source ) ):
+			$source = 'inbox-notes';
+		elseif ( preg_match( '/\Sadmin-ajax.php/', $source ) ):
+			$source = 'heartbeat-api';
+		endif;
```
I wonder if `stripos` would work for some or all of these conditions? `preg_match` is exact, but a bit slow, and we're doing it several times here.

KokkieH replied:
It worked perfectly. Thanks for the suggestion!

andfinally commented on plugins/woocommerce/includes/admin/helper/class-wc-helper.php:
```diff
@@ -1270,11 +1270,24 @@ public static function get_subscriptions() {
 			'subscriptions',
 			array(
 				'authenticated' => true,
+				'source'        => '?source=' . $source,
```
I reckon we should escape this, just to be safe. It might make sense to name the arg something more generic here, in `WC_Helper_API::request` and `WC_Helper_API::url`, as it's possible someone will want to add other URL params at some point. For example:

```suggestion
				'query_string'        => esc_url( '?source=' . $source ),
```

KokkieH replied:
Thanks for the suggestions. Makes sense making it more generic.

andfinally reviewed (commented):
Looking good! 👌 Just left some minor thoughts.

KokkieH committed:
Use sripos instead of preg_match in WC_Helper::get_subscriptions()

- Update "source" argument name across methods to "query_string" to be
  more generic
- Use separate variables to make it clear the URI of the requesting
  page, and the value passed as the source query string are different

jorgeatorres reviewed (dismissed):
LGTM 👍

andfinally commented on plugins/woocommerce/includes/admin/helper/class-wc-helper-api.php:
```diff
@@ -166,10 +166,11 @@ public static function put( $endpoint, $args = array() ) {
 	 * @return string The absolute endpoint URL.
 	 */
-	public static function url( $endpoint, $source = '' ) {
+	public static function url( $endpoint, $query_string = '' ) {
```
We should declare `$query_string` as a `@param` in the function comment.

KokkieH replied:
Thanks! Added in https://github.com/woocommerce/woocommerce/pull/35051/commits/f31d9d9dbd5e4980876efb42e76ee5c547557f8f

andfinally commented:
I'm just having a go at testing this, but I'm having trouble checking it out into my local WooCommerce root. I get `pathspec 'KokkieH:add/source-param-to-get-subscriptions-call' did not match any file(s) known to git`.

KokkieH commented:
@andfinally could it be because I created the PR via a fork of the repo?

KokkieH commented:
@andfinally here's a [patch of the changes](https://github.com/woocommerce/woocommerce/files/9817538/core-helper-patch.patch.txt) (minus the changelog update).

andfinally commented on plugins/woocommerce/includes/admin/helper/class-wc-helper.php:
```diff
@@ -1270,11 +1270,25 @@ public static function get_subscriptions() {
 			array(
 				'authenticated' => true,
+				'query_string'  => esc_url( '?source=' . $source ),
+
```
We can delete this stray empty line.

KokkieH replied:
Ah, thanks for catching that.

andfinally reviewed (dismissed):
Looks good to me, thanks for the help with testing. Confirmed that `source=my_subscriptions` is added to the URL in the extensions tab request, and that other requests still work as expected. 

(The `active/inactive` switch on the My Subscriptions tab isn't working on my customer test site, but it never has, so I don't think it's related. On wp-env, the switch isn't shown, perhaps because the site URL is http.)

KokkieH committed:
Add @param declaration to function comment

KokkieH commented:
@jorgeatorres it appears my last changes based on @andfinally 's feedback invalidated you original review. Can you please check and approve this again?

There are also two failing CI checks. I looked through them - I'm not sure what needs to be done for the first, and the second doesn't seem related to any of the changes I made. Can you please take a look as well? Thanks!

jorgeatorres committed:
Address PHPCS issues

jorgeatorres reviewed (approved):
LGTM.
